src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1305,17 +1346,27 @@ uct_dc_mlx5_iface_dci_do_dcs_pending_tx(ucs_arbiter_t *arbiter,
     int is_only                = ucs_arbiter_elem_is_only(elem);
     ucs_arbiter_cb_result_t res;
 
+    ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
+
     res = uct_dc_mlx5_iface_dci_do_common_pending_tx(ep, elem);
-    if (res == UCS_ARBITER_CB_RESULT_REMOVE_ELEM) {
-        /* For dcs* policies release dci if this is the last elem in the group
-         * and the dci has no outstanding operations. For example pending
-         * callback did not send anything. (uct_ep_flush or just return ok)
-         */
-        if (is_only) {
-            uct_dc_mlx5_iface_dci_free(iface, ep);
-        }
+    if ((res != UCS_ARBITER_CB_RESULT_REMOVE_ELEM) || !is_only) {
+        return res;
+    }
+
+    /* For dcs* policies release dci if this is the last elem in the group
+     * and the dci has no outstanding operations. For example pending
+     * callback did not send anything. (uct_ep_flush or just return ok)
+     */
+    if (!uct_dc_mlx5_iface_dci_detach(iface, ep,
+                                      UCT_DC_IFACE_DCI_EP_DESTROYED)) {
```

Comment:
maybe rename this macro to UCT_DC_IFACE_DCI_EP_DETACHED ?

Reply:
renamed to UCT_DC_IFACE_DCI_EP_DETACHED